ci/linux/run.sh

```diff
@@ -0,0 +1,295 @@
+        '--all'|'--no-all')
+            [[ "$1" = --no-* ]]
+            VALUE=$?
+            CLEAN=$VALUE
+            NPM_BUILD=$VALUE
+            SLN_RESTORE=$VALUE
+            SLN_BUILD=$VALUE
+            UNIT_TESTS=$VALUE
+            JS_TESTS=$VALUE
+            UI_TESTS=$VALUE
+            shift
+            continue
+        ;;
+        '--root')
+            DOTVVM_ROOT="$2"
+            shift 2
+            continue
+        ;;
+        '--config')
+            BUILD_CONFIGURATION="$2"
+            shift 2
+            continue
+        ;;
+        '--samples-profile')
+            SAMPLES_PROFILE="$2"
+            shift 2
+            continue
+        ;;
+        '--samples-port')
+            SAMPLES_PORT="$2"
+            shift 2
+            continue
+        ;;
+        '--samples-port-api')
+            SAMPLES_PORT_API="$2"
+            shift 2
+            continue
+        ;;
+        '--')
+            shift
+            break
+        ;;
+        *)
+            # handle all flags
+            [[ "$1" = --no-* ]]
+            VALUE=$?
+            if [ $VALUE -eq 1 ]; then
+                OPTION="${1#--}";
+            else
+                OPTION="${1#--no-}";
+            fi
+
+            OPTION=${OPTION^^}
+            OPTION=${OPTION//-/_}
+            eval test -n "\$$OPTION"
+            if [ $? -ne 0 ]; then
+                # this flag doesn't exit
+                echo >&2 "Option '$1' is not recognized."
+                exit 1
+            fi
+
+            eval $OPTION=$VALUE
+            shift
+            continue;
+        ;;
+    esac
+done
+
+if [ -n "$1" ]; then
+    echo >&2 "Argument '$1' is invalid."
+    exit 1
+fi
+
+# ==================
+# config var setting
+# ==================
+
+ROOT=${DOTVVM_ROOT:-$(pwd)}
+# override DOTVVM_ROOT in case this is a local build
+export DOTVVM_ROOT=$ROOT
+
+CONFIGURATION="${BUILD_CONFIGURATION:-Release}"
+DISPLAY="${DISPLAY:-":42"}"
+export DISPLAY
+SLN="$ROOT/ci/linux/Linux.sln"
+TEST_RESULTS_DIR="$ROOT/artifacts/test"
+SAMPLES_DIR="$ROOT/src/DotVVM.Samples.Tests"
+SAMPLES_PROFILE="${SAMPLES_PROFILE:-seleniumconfig.aspnetcorelatest.chrome.json}"
+SAMPLES_PORT="${SAMPLES_PORT:-16019}"
+SAMPLES_PORT_API="${SAMPLES_PORT_API:-5001}"
+
+tput sgr0
+echo "ROOT=$ROOT"
+echo "SLN=$SLN"
+echo "CONFIGURATION=$CONFIGURATION"
+echo "DISPLAY=$DISPLAY"
+echo "TEST_RESULTS_DIR=$TEST_RESULTS_DIR"
+echo "SAMPLES_DIR=$SAMPLES_DIR"
+echo "SAMPLES_PROFILE=$SAMPLES_PROFILE"
+echo "SAMPLES_PORT=$SAMPLES_PORT"
+echo "SAMPLES_PORT_API=$SAMPLES_PORT_API"
+
+# ================
+# helper functions
+# ================
+
+function print_header {
+    tput sgr0
+    echo "$(tput setaf 3)--------------------------------"
+    echo "$(tput setaf 3 && tput smso)$@$(tput rmso)"
+    echo "$(tput setaf 3)--------------------------------"
+    tput sgr0
+}
+
+function run_named_command {
+    NAME=$1
+    shift
+
+    print_header $NAME
+    tput setaf 4 && printf "running '$@'\n" && tput sgr0
```

Comment:
Why use `printf` instead of `echo`? Also it would be safer to write something like `printf "running %s\n" "$@"`

Reply:
I don't recall why I decided to use `printf` but `echo` works just as nicely.